src/leap/common/service_hooks.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+# service_hooks.py
+# Copyright (C) 2016 LEAP
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Hooks for service composition.
+"""
+from collections import defaultdict
+
+from twisted.python import log
+
+
+class HookableService(object):
+
+    """
+    This service allows for other services in a Twisted Service tree to be
+    notified whenever a certain kind of hook is triggered.
+
+    During the service composition, one is expected to register
+    a hook name with the service that wants to react to the triggering of
+    the hook. The listener service must add this hook name to a class variable
+    called 'subscribed_to_hooks'.
+
+    Upon the hook being triggered, the method "hook_<name>" will be called with
+    the passed data in the listener service.
+    """
+
+    def register_hook(self, name, listener):
+        if not hasattr(self, 'event_listeners'):
+            self.event_listeners = defaultdict(list)
+        log.msg("Registering hook %s->%s" % (name, listener))
+        self.event_listeners[name].append(listener)
+
+    def trigger_hook(self, name, **data):
+
+        def react_to_hook(listener, name, **kw):
+            if name not in listener.subscribed_to_hooks:
```

Comment:
I have the feeling that the user of `HoockableService` needs to pay attention to too many details to use it. Can we simplify it somehow?